Dockerfile

```diff
@@ -13,35 +13,59 @@ ENV WOLFRAM_APP_ID=$wolfram_app_id
 # install system dependencies
 RUN apt-get update
 RUN apt-get -y install ca-certificates
-RUN apt-get -y install python3 python3-pip python3-venv
-RUN apt-get -y install curl gnupg
+RUN apt-get -y install python3 python3-pip python3-venv 
+RUN apt-get -y install curl gnupg 
 RUN apt-get -y install libpq-dev libgmp-dev libgmp3-dev libmpfr-dev libmpc-dev
 RUN apt-get install -y npm
 RUN npm i -g n && n lts && npm i -g npm@latest
 
-COPY ./react-frontend ./ 
-
-# install node package dependencies
+# React frontend
+COPY react-frontend ./react-frontend
+WORKDIR /srv/ramanujan-machine-web-portal/react-frontend
+# install NodeJS dependencies
 RUN npm install
-
-# Build frontend project
+# generate deployable artifacts
 RUN npm run build
 
-# copy API files
-COPY ./python-backend/Pipfile* ./
-COPY ./python-backend/*.py ./
-COPY ./python-backend/requirements.txt ./
+# create virtual environment for FastAPI web server and install dependencies
+WORKDIR /srv/ramanujan-machine-web-portal
+ENV BKPATH=$PATH
+COPY python-backend ./python-backend
+RUN cp -rf react-frontend/build python-backend/build
+WORKDIR /srv/ramanujan-machine-web-portal/python-backend
+RUN rm -rf venv
+RUN python3 -m venv venv
+RUN . venv/bin/activate
+RUN ./venv/bin/python -m pip install uvicorn
+RUN ./venv/bin/python -m pip install -r /srv/ramanujan-machine-web-portal/python-backend/requirements.txt
+ENV PATH=$BKPATH
+RUN unset VIRTUAL_ENV
 
-# install Python dependencies
+# create virtual environment for gRPC server and install dependencies
+WORKDIR /srv/ramanujan-machine-web-portal
+COPY protos ./protos
+COPY lirec-grpc-server ./lirec-grpc-server
+WORKDIR /srv/ramanujan-machine-web-portal/lirec-grpc-server
+RUN rm -rf venv
 RUN python3 -m venv venv
-RUN chmod +x venv/bin/activate
-RUN ./venv/bin/activate
-RUN pip3 install -r requirements.txt --break-system-packages
-RUN pip3 install uvicorn --break-system-packages
+RUN . venv/bin/activate
+RUN venv/bin/python -m pip install -r requirements.txt
+ENV PATH=$BKPATH
+RUN unset VIRTUAL_ENV
+
+# generate gRPC code in both client (web server) and gRPC server directories
+WORKDIR /srv/ramanujan-machine-web-portal
+RUN pip install grpcio grpcio-tools --break-system-packages
+WORKDIR /srv/ramanujan-machine-web-portal/python-backend
+RUN python3 -m grpc_tools.protoc --proto_path=/srv/ramanujan-machine-web-portal/protos --python_out=/srv/ramanujan-machine-web-portal/python-backend/ --grpc_python_out=/srv/ramanujan-machine-web-portal/python-backend/ /srv/ramanujan-machine-web-portal/protos/lirec.proto
```

Comment:
Should this activity be done during the build rather than during image generation?